src/domain.jl

```diff
@@ -0,0 +1,172 @@
+
+    # define array of next time step, absolute tolerance, and scale factor
+    u = typeof(f.u) <: Void ? similar(integrator.u) : f.u
+    abstol = typeof(f.abstol) <: Void ? integrator.opts.abstol : f.abstol
+    scalefactor = typeof(f.scalefactor) <: Void ? integrator.opts.qmin : f.scalefactor
+
+    # setup callback and save addtional arguments for checking next time step
+    args = setup(f, integrator)
+
+    # cache current time step
+    dt = integrator.dt
+
+    # update time step of integrator to proposed next time step
+    integrator.dt = get_proposed_dt(integrator)
+
+    # adjust time step to bounds and time stops
+    fix_dt_at_bounds!(integrator)
+    modify_dt_for_tstops!(integrator)
+
+    while integrator.tdir * integrator.dt > 0
+        # calculate estimated value of next step and its residuals
+        integrator(u, integrator.t + integrator.dt)
+
+        # check whether time step is accepted
+        isaccepted(u, abstol, f, args...) && break
+
+        # adjust time step
+        dtcache = integrator.dt
+        integrator.dt *= scalefactor
+        fix_dt_at_bounds!(integrator)
+        modify_dt_for_tstops!(integrator)
+
+        # abort iteration when time step is not changed
+        if dtcache == integrator.dt
+            warn("Could not restrict values to domain. Iteration was canceled since ",
+                 "time step dt = ", integrator.dt, " could not be reduced.")
+            break
+        end
+    end
+
+    # update current and next time step
+    set_proposed_dt!(integrator, integrator.dt)
+    integrator.dt = dt
+end
+
+"""
+    modify_u!(integrator, f::AbstractDomainAffect)
+
+Modify current state vector `u` of `integrator` if required, and return whether it actually
+was modified.
+"""
+modify_u!(integrator, ::AbstractDomainAffect) = false
+
+"""
+    setup(f::AbstractDomainAffect, integrator)
+
+Setup callback `f` and return an arbitrary tuple whose elements are used as additional
+arguments in checking whether time step is accepted.
+"""
+setup(::AbstractDomainAffect, integrator) = ()
+
+"""
+    isaccepted(u, abstol, f::AbstractDomainAffect, args...)
+
+Return whether `u` is an acceptable state vector at the next time point given absolute
+tolerance `abstol`, callback `f`, and other optional arguments.
+"""
+isaccepted(u, tolerance, ::AbstractDomainAffect, args...) = true
+
+# specific method definitions for positive domain callback
+
+function modify_u!(integrator, f::PositiveDomainAffect)
+    modified = false
+
+    # set all negative values to zero
+    @inbounds for i in eachindex(integrator.u)
+        if integrator.u[i] < 0
+            integrator.u[i] = 0
+            modified = true
+        end
+    end
+
+    modified
+end
+
+# state vector is accepted if its entries are greater than -abstol
+isaccepted(u, abstol::Number, ::PositiveDomainAffect) = all(x -> x + abstol > 0, u)
+isaccepted(u, abstol, ::PositiveDomainAffect) = all(x + y > 0 for (x,y) in zip(u, abstol))
```

Comment:
Okay just checking. Will this not cause problems when non-negativity is is mathematically necessary, like chemical reactions which would be exponential if negative?